### PR TITLE
Implement upgrade client proposal handler

### LINKED ISSUE #141: Handler for upgrade client
## Crate

`ibc`

## Summary

- [x] handler implementation in ics02_client::handler::upgrade_client.rs
     - raise an issue with cosmos/ics to add canonical algorithm for handling this message? not clear if that's desirable since this is documented and handled mostly in the SDK: https://github.com/cosmos/cosmos-sdk/tree/master/docs/ibc/upgrades
     - probalby inspired from https://github.com/cosmos/cosmos-sdk/pull/7367 since there is no canonical algorithm in the cosmos/ics spec (yet)
- [ ] clarify the use of `upgrade_path` field in [Tendermint client state](https://github.com/informalsystems/ibc-rs/blob/db2be6f72fef647933f48d80b29a0f9f12c804d4/modules/src/ics07_tendermint/client_state.rs#L25) and include that in tests

____

#### For Admin Use

- [X] Not duplicate issue
- [X] Appropriate labels applied
- [X] Appropriate milestone (priority) applied
- [X] Appropriate contributors tagged
- [X] Contributor assigned/self-assigned

### LINKED ISSUE #420: Provide a "upgrade client proposal" handler for Tendermint chains
## Problem Definition

The work done so far and covered within [ADR06](https://github.com/cosmos/ibc-rs/blob/main/docs/architecture/adr-006-upgrade-client-implementation.md) around upgrading a counterparty client mainly reflects how to handle submitted `MsgUpgradeClient` by the relayer and substitute the upgraded client and consensus states on the counterparty chain. 
But, there is another side to this process (see [ibc-go](https://github.com/cosmos/ibc-go/blob/e093d85b533ab3572b32a7de60b88a0816bed4af/modules/core/02-client/keeper/proposal.go#L71-L102)) and it is where the newly proposed client and consensus states should be stored in the _upgrade path_ (related to #141) of the upgrading chain. 
We haven't clarified yet how the incoming chain-level upgrade proposal should be handled within the IBC-rs core and a specific implementation for Tendermint chains seems to be necessary.

### PR DESCRIPTION
**Pull request by Farhad-Shabani**
Closes: #141
Closes: #420

See sample implementation for basecoin-rs: https://github.com/informalsystems/basecoin-rs/pull/101

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests.
- [X] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: crates/ibc/src/hosts/tendermint/upgrade_proposal/mod.rs
Comment (plafer) [inline crates/ibc/src/hosts/tendermint/upgrade_proposal/mod.rs]: These types are missing a docstring

File: crates/ibc/src/hosts/tendermint/upgrade_proposal/context.rs
Comment (plafer) [inline crates/ibc/src/hosts/tendermint/upgrade_proposal/context.rs:18]: I now think it makes more sense to bring `upgrade_client_proposal_handler()` (now defined in basecoin) back in ibc-rs. Otherwise, we have these traits defined, and nothing that uses them. And I now understand that this handler/traits are optional; chains are always free to implement upgrades however they want. So we can start with the `upgrade_client_proposal_handler` we currently have, and iteratively improve it over time

**Review by plafer — COMMENTED** [review]
Looks great!

File: crates/ibc/src/hosts/tendermint/upgrade_proposal/mod.rs
Comment (Farhad-Shabani) [reply crates/ibc/src/hosts/tendermint/upgrade_proposal/mod.rs]: [c711d40](https://github.com/cosmos/ibc-rs/pull/680/commits/c711d40ef7d624022c2dcdbf5c0c1579eb1734dc)

File: crates/ibc/src/hosts/tendermint/upgrade_proposal/context.rs
Comment (Farhad-Shabani) [reply crates/ibc/src/hosts/tendermint/upgrade_proposal/context.rs:18]: This was chain-dependent at some point while implementing, but with the latest code, right, it can now be placed in IBC-rs.
[2950ab7](https://github.com/cosmos/ibc-rs/pull/680/commits/2950ab7232f7d0514204334fa772faf6c4a43da4)

**Comment by plafer** [comment]
Latest changes look good, I'll wait for the basecoin-rs PR to be ready too before we merge this

**Review by plafer — APPROVED** [review]
👌

**Pull request merged**
